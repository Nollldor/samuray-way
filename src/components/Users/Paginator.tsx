import React, {FC} from "react";
import styles from "./paginator.module.css"

type PropsType = {
    totalUsersCount: number
    pageSize: number
    currentPage: number
    onPageChanged: (page: number) => void
}

export const Paginator: FC<PropsType> = (
    {
        totalUsersCount,
        pageSize,
        currentPage,
        onPageChanged
    }) => {
    const pageCount = Math.ceil(totalUsersCount / pageSize)
    let pages = []
    for (let i = 1; i <= pageCount; i++) {
        pages.push(i)
    }

    return <div>
        {pages.map((p, i) => <span className={currentPage === p ? styles.selectedPage : ""} key={i}
                                   onClick={() => onPageChanged(p)}>{p}</span>)}
    </div>
}