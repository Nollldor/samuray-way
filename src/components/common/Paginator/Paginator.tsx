import React, {FC, useState} from "react";
import styles from "components/common/Paginator/Paginator.module.css"
import cn from 'classnames'

type PropsType = {
    totalItemsCount: number
    pageSize: number
    currentPage: number
    onPageChanged: (page: number) => void
    portionSize: number
}

export const Paginator: FC<PropsType> = (
    {
        totalItemsCount,
        pageSize,
        currentPage,
        onPageChanged,
        portionSize
    }) => {
    const pageCount = Math.ceil(totalItemsCount / pageSize)
    let pages = []
    for (let i = 1; i <= pageCount; i++) {
        pages.push(i)
    }

    const portionCount = Math.ceil(pageCount / portionSize)
    const [portionNumber, setPortionNumber] = useState(1)
    const leftPortionPageNumber = (portionNumber - 1) * portionSize + 1
    const rightPortionPageNumber = portionNumber * portionSize

    return <div className={styles.paginator}>
        {portionNumber > 1 && <button onClick={() => {
            setPortionNumber(portionNumber - 1)
        }}>Prev</button>}
        {pages
            .filter(p => p >= leftPortionPageNumber && p <= rightPortionPageNumber)
            .map((p) => {
                return <span className={cn({[styles.selectedPage]: currentPage === p}, styles.pageNumber)} key={p}
                             onClick={(e) => {
                                 onPageChanged(p)
                             }}
                >{p}</span>
            })}
        {portionCount > portionNumber && <button onClick={() => setPortionNumber(portionNumber + 1)}>Next</button>}
    </div>
}