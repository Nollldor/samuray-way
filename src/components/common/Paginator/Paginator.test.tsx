import React from "react";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";
import { Paginator, PaginatorPropsType } from "./Paginator";

describe("Paginator component", () => {
    const onPageChangedMock = jest.fn();

    it("renders correctly with current page selected", () => {
        const component = renderer.create(
            <Paginator
                totalItemsCount={10}
                pageSize={5}
                currentPage={2}
                onPageChanged={onPageChangedMock}
                portionSize={3}
            />
        );

        const tree = component.toJSON() as ReactTestRendererJSON;
        expect(tree).toMatchSnapshot();
    });

    it("calls onPageChanged when a page is clicked", () => {
        const component = renderer.create(
            <Paginator
                totalItemsCount={10}
                pageSize={5}
                currentPage={2}
                onPageChanged={onPageChangedMock}
                portionSize={3}
            />
        );

        const pageElements = component.root.findAllByType("span");
        const secondPageElement = pageElements.find(
            (element) => element.props.children === 2
        );

        if (secondPageElement) {
            secondPageElement.props.onClick();
        }

        expect(onPageChangedMock).toHaveBeenCalledWith(2);
    });
});
