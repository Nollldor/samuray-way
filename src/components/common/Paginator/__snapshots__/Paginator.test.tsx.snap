// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator component renders correctly with current page selected 1`] = `
<div
  className="paginator"
>
  <span
    className="pageNumber"
    onClick={[Function]}
  >
    1
  </span>
  <span
    className="selectedPage pageNumber"
    onClick={[Function]}
  >
    2
  </span>
</div>
`;
